Release/src/http/client/http_client_curl.cpp

```diff
@@ -0,0 +1,2379 @@
+
+        std::vector<uint8_t> m_buffer; // we read data from the stream into this before compressing
+        uint8_t* m_acquired; // we use this in place of m_buffer if the stream has directly-accessible data available
+        size_t m_bytes_read; // we most recently read this many bytes, which may be less than m_buffer.size()
+        size_t m_bytes_processed; // we've compressed this many bytes of m_bytes_read so far
+        bool m_needs_flush;   // we've read and compressed all bytes, but the compressor still has compressed bytes to
+                              // give us
+        bool m_started;       // we've sent at least some number of bytes to m_decompressor
+        bool m_done;          // we've read, compressed, and consumed all bytes
+        bool m_chunked;       // if true, we need to decode and decompress a transfer-encoded message
+        size_t m_chunk_bytes; // un-decompressed bytes remaining in the most-recently-obtained data from m_chunk
+        std::unique_ptr<_chunk_helper> m_chunk;
+    } m_compression_state;
+
+    void cleanup()
+    {
+        if (m_compression_state.m_acquired != nullptr)
+        {
+            // We may still hold a piece of the buffer if we encountered an exception; release it here
+            if (m_decompressor)
+            {
+                _get_writebuffer().commit(0);
+            }
+            else
+            {
+                _get_readbuffer().release(m_compression_state.m_acquired, m_compression_state.m_bytes_processed);
+            }
+            m_compression_state.m_acquired = nullptr;
+        }
+
+        if (m_request_handle != nullptr)
+        {
+            WinHttpCloseHandle(m_request_handle);
+        }
+    }
+
+    void install_custom_cn_check(const utility::string_t& customHost)
+    {
+        m_customCnCheck = customHost;
+        utility::details::inplace_tolower(m_customCnCheck);
+    }
+
+    void on_send_request_validate_cn()
+    {
+        // we do the validation inside curl
+    }
+
+protected:
+    virtual void finish() override
+    {
+        request_context::finish();
+        assert(m_self_reference != nullptr);
+        auto dereference_self = std::move(m_self_reference);
+        // As the stack frame cleans up, this will be deleted if no other references exist.
+    }
+
+private:
+    utility::string_t m_customCnCheck;
+    std::vector<unsigned char> m_cachedEncodedCert;
+
+    // Can only create on the heap using factory function.
+    winhttp_request_context(const std::shared_ptr<_http_client_communicator>& client, const http_request& request)
+        : request_context(client, request)
+        , m_request_handle(nullptr)
+        , m_proxy_authentication_tried(false)
+        , m_server_authentication_tried(false)
+        , m_bodyType(no_body)
+        , m_remaining_to_write(0)
+        , m_startingPosition(std::char_traits<uint8_t>::eof())
+        , m_readStream(request.body())
+        , m_body_data()
+    {
+    }
+};
+
+
+#if 0
```

Comment:
Please don't check in commented out code.

Reply:
removed